gtsfm/frontend/detector_descriptor/d2net.py

```diff
@@ -0,0 +1,125 @@
+from thirdparty.d2net.lib.model_test import D2Net
+from thirdparty.d2net.lib.utils import preprocess_image
+
+MAX_EDGE_PX = 1600
+MAX_SUM_EDGES_PX = 2800
+USE_MULTISCALE = False
+USE_RELU = True
+PREPROCESSING_METHOD = "torch"
+PYRAMID_SCALES = [0.5, 1, 2]
+
+MODEL_PATH = Path(__file__).resolve().parent.parent.parent.parent / "thirdparty" / "d2net" / "weights" / "d2_tf.pth"
+
+
+class D2NetDetDesc(DetectorDescriptorBase):
+    """D2-Net detector descriptor."""
+
+    def __init__(self, max_keypoints: int = 5000, model_path: Path = MODEL_PATH, use_cuda: bool = True) -> None:
+        """Instantiate parameters and hardware settings for D2-Net detector-descriptor.
+
+        We set the maximum number of keypoints, set the path to pre-trained weights, and determine whether
+        CUDA enabled devices can be utilized for inference.
+        """
+        super().__init__()
+        self.max_keypoints = max_keypoints
+        self.model_path = model_path
+        self.use_cuda = use_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if self.use_cuda else "cpu")
+
+    def detect_and_describe(self, image: Image) -> Tuple[Keypoints, np.ndarray]:
+        """Extract keypoints and their corresponding descriptors.
+
+        Adapted from:
+        https://github.com/mihaidusmanu/d2-net/blob/master/extract_features.py
+
+        Args:
+            image: the input image.
+
+        Returns:
+            Detected keypoints, with length N <= max_keypoints.
+            Corr. descriptors, of shape (N, D) where D is the dimension of each descriptor.
+        """
+        model = D2Net(model_file=self.model_path, use_relu=USE_RELU, use_cuda=self.use_cuda)
+        model.eval()
+
+        # Resize image, and obtain re-scaling factors to postprocess keypoint coordinates.
+        resized_image, fact_i, fact_j = resize_image(image.value_array)
+        input_image = preprocess_image(resized_image, preprocessing=PREPROCESSING_METHOD)
+
+        if USE_MULTISCALE:
+            scales = PYRAMID_SCALES
+        else:
+            scales = [1]
+
+        with torch.no_grad():
+            keypoints, scores, descriptors = d2net_pyramid.process_multiscale(
+                image=torch.tensor(input_image[np.newaxis, :, :, :].astype(np.float32), device=self.device),
+                model=model,
+                scales=scales,
+            )
+
+        # Choose the top K keypoints and descriptors.
+        ordered_idxs = np.argsort(-scores)[: self.max_keypoints]
+        keypoints = keypoints[ordered_idxs, :]
+        descriptors = descriptors[ordered_idxs, :]
+
+        # Rescale keypoint coordinates from resized image scale, back to provided input image resolution.
+        keypoints[:, 0] *= fact_i
+        keypoints[:, 1] *= fact_j
+
+        # Convert (y,x) tuples that represented (i, j) indices of image matrix, into (u, v) coordinates.
+        keypoints = keypoints[:, [1, 0]]
+        return Keypoints(coordinates=keypoints), descriptors
```

Comment:
Can we add score information in the keypoints?

Reply:
Sure, just added.